src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs

```diff
@@ -8350,6 +8371,755 @@ class C
+        {
+            var comp = CreateCompilation(new[] { @"
+using System.Runtime.InteropServices;
+public struct S
+{
+    public static void M2(S s)
+    {
+        foreach (var i in s) {}
+    }
+}
+public ref struct SEnumerator
+{
+    public bool MoveNext() => throw null;
+    public int Current => throw null;
+    [UnmanagedCallersOnly]
+    public void Dispose() => throw null;
+}
+public static class CExt
+{
+    public static SEnumerator GetEnumerator(this S s) => throw null;
+}
+", UnmanagedCallersOnlyAttribute });
+
+            comp.VerifyDiagnostics(
+                // (14,6): error CS8896: 'UnmanagedCallersOnly' can only be applied to ordinary static methods or static local functions.
+                //     [UnmanagedCallersOnly]
+                Diagnostic(ErrorCode.ERR_UnmanagedCallersOnlyRequiresStatic, "UnmanagedCallersOnly").WithLocation(14, 6)
+            );
+        }
+
+        [Fact]
+        public void UnmanagedCallersOnlyCannotCaptureToDelegate()
+        {
+            var comp = CreateCompilation(new[] { @"
+using System;
+using System.Runtime.InteropServices;
+public class C
+{
+    [UnmanagedCallersOnly]
+    public static void M1() { }
+
+    public static void M2()
+    {
+        Action a = M1;
+        a = local;
+        a = new Action(M1);
+        a = new Action(local);
+
+        [UnmanagedCallersOnly]
+        static void local() {}
+    }
+}
+", UnmanagedCallersOnlyAttribute });
+
+            comp.VerifyDiagnostics(
+                // (11,20): error CS8897: 'C.M1()' is attributed with 'UnmanagedCallersOnly' cannot be called directly. Obtain a function pointer to this method.
+                //         Action a = M1;
+                Diagnostic(ErrorCode.ERR_UnmanagedCallersOnlyMethodsCannotBeCalledDirectly, "M1").WithArguments("C.M1()").WithLocation(11, 20),
+                // (12,13): error CS8897: 'local()' is attributed with 'UnmanagedCallersOnly' cannot be called directly. Obtain a function pointer to this method.
+                //         a = local;
+                Diagnostic(ErrorCode.ERR_UnmanagedCallersOnlyMethodsCannotBeCalledDirectly, "local").WithArguments("local()").WithLocation(12, 13),
+                // (13,24): error CS8897: 'C.M1()' is attributed with 'UnmanagedCallersOnly' cannot be called directly. Obtain a function pointer to this method.
+                //         a = new Action(M1);
+                Diagnostic(ErrorCode.ERR_UnmanagedCallersOnlyMethodsCannotBeCalledDirectly, "M1").WithArguments("C.M1()").WithLocation(13, 24),
+                // (14,24): error CS8897: 'local()' is attributed with 'UnmanagedCallersOnly' cannot be called directly. Obtain a function pointer to this method.
+                //         a = new Action(local);
+                Diagnostic(ErrorCode.ERR_UnmanagedCallersOnlyMethodsCannotBeCalledDirectly, "local").WithArguments("local()").WithLocation(14, 24)
+            );
+        }
```

Comment:
>} [](start = 8, length = 1)

Another test suggestion to show that error on delegate conversion doesn't affect selection of delegate:
```
using System;
public class C 
{
    [UnmanagedCallersOnly]
    public static void M(string s) { }
    
    public static void M(object o) { }
    
    void N()
    {
        Action<string> a = M;   // still picks M(string) and errors, so attribute doesn't affect applicability (same as Obsolete)
    }
}
```